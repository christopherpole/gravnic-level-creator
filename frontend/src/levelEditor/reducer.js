import availableTiles from 'config/tiles';
import { SAVE_LEVEL, LOAD_LEVEL_CONFIRMED } from 'levelManager/actions';
import { createNewLevel } from 'utils';
import { MIN_MOVES, MAX_MOVES } from 'config/settings';
import {
  UPDATE_TILE,
  SELECT_TILE,
  RESET_GRID,
  START_DRAG,
  STOP_DRAG,
  SET_STARS,
  SET_LINK_TO_TILE_POS,
  CREATE_LINK,
} from './actions';

const newLevel = createNewLevel();

export const initialState = {
  previewing: false,
  dragging: false,
  linkFromTilePos: null,
  linkToTilePos: null,
  selectedTileId: availableTiles[0].id,
  availableTiles,
  tiles: newLevel.tiles,
  stars: newLevel.stars,
  editedSinceLastSave: false,
  links: [],
};

export default function levelEditorReducer(state = initialState, action) {
  switch (action.type) {
    case SELECT_TILE: {
      return {
        ...state,
        selectedTileId: action.selectedTileId,
      };
    }

    case UPDATE_TILE: {
      const newTiles = state.tiles.slice();
      newTiles[action.position] = {
        ...newTiles[action.position],
        selectedTileId: state.selectedTileId,
      };

      const newLinks = state.links.filter(
        link => link.from !== action.position && link.to !== action.position,
      );

      return {
        ...state,
        tiles: newTiles,
        editedSinceLastSave: true,
        links: newLinks,
      };
    }

    case RESET_GRID: {
      return {
        ...state,
        tiles: initialState.tiles,
        editedSinceLastSave: false,
        links: initialState.links,
      };
    }

    case SAVE_LEVEL: {
      return {
        ...state,
        editedSinceLastSave: false,
      };
    }

    case LOAD_LEVEL_CONFIRMED: {
      return {
        ...state,
        tiles: action.level.tiles,
        stars: action.level.stars,
        editedSinceLastSave: false,
        links: action.level.links,
      };
    }

    case START_DRAG: {
      return {
        ...state,
        dragging: true,
        linkFromTilePos: action.linkFromTilePos,
      };
    }

    case STOP_DRAG: {
      const links = [...state.links];

      return {
        ...state,
        dragging: false,
        linkFromTilePos: null,
        linkToTilePos: null,
        links,
      };
    }

    case SET_LINK_TO_TILE_POS: {
      return {
        ...state,
        dragging: false,
        linkToTilePos: action.position,
      };
    }

    case CREATE_LINK: {
      return {
        ...state,
        linkFromTilePos: state.linkFromTilePos,
        linkToTilePos: state.linkToTilePos,
        links: [...state.links, { from: state.linkFromTilePos, to: state.linkToTilePos }],
      };
    }

    case SET_STARS: {
      const first = [...state.stars.slice(0, action.starsIndex)].map(noOfMoves =>
        Math.min(noOfMoves, action.stars),
      );

      const rest = [...state.stars.slice(action.starsIndex + 1)].map(noOfMoves =>
        Math.max(noOfMoves, action.stars),
      );

      //  Ensure values are between 1 and 99
      const arr = [...first, action.stars, ...rest].map(val => {
        if (val < MIN_MOVES) return MIN_MOVES;
        if (val > MAX_MOVES) return MAX_MOVES;
        return val;
      });

      return {
        ...state,
        stars: arr,
        editedSinceLastSave: true,
      };
    }

    default: {
      return state;
    }
  }
}
