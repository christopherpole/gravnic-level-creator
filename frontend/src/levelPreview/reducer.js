import { cloneDeep } from 'lodash';
import { GAME_SPEED_NORMAL } from 'config/settings';

import {
  PREVIEW_LEVEL,
  EDIT_LEVEL,
  CHANGE_GRAVITY_DIRECTION,
  UPDATE_GAME_STATE,
  ENTITIES_STOPPED_MOVING,
  RESTART_LEVEL,
  UNDO_MOVE_STEP,
  UNDO_MOVE_FINISHED,
  SET_GAME_SPEED,
} from './actions';

export const initialState = {
  previewing: false,
  gameState: null,
  gravityDirection: null,
  entitiesMoving: false,
  gameHistory: [],
  gameSpeed: GAME_SPEED_NORMAL,
};

export default function levelPreviewReducer(state = initialState, action) {
  switch (action.type) {
    case PREVIEW_LEVEL: {
      return {
        ...state,
        previewing: true,
        gameState: action.gameState,
        gameHistory: [action.gameState],
        entitiesMoving: initialState.entitiesMoving,
        gravityDirection: initialState.gravityDirection,
      };
    }

    case EDIT_LEVEL: {
      return {
        ...state,
        previewing: false,
      };
    }

    case CHANGE_GRAVITY_DIRECTION: {
      const newGameHistory = cloneDeep(state.gameHistory);
      newGameHistory.push([cloneDeep(state.gameState)]);

      return {
        ...state,
        gravityDirection: action.direction,
        entitiesMoving: true,
        gameHistory: newGameHistory,
      };
    }

    case UPDATE_GAME_STATE: {
      const newGameHistory = cloneDeep(state.gameHistory);
      newGameHistory[newGameHistory.length - 1].push(action.gameState);

      return {
        ...state,
        gameState: action.gameState,
        gameHistory: newGameHistory,
      };
    }

    case ENTITIES_STOPPED_MOVING: {
      return {
        ...state,
        entitiesMoving: false,
      };
    }

    case RESTART_LEVEL: {
      return {
        ...state,
        entitiesMoving: initialState.entitiesMoving,
        gameHistory: [cloneDeep(state.gameHistory[0])],
        gravityDirection: initialState.gravityDirection,
        gameState: cloneDeep(state.gameHistory[0]),
      };
    }

    case UNDO_MOVE_STEP: {
      const gameHistoryLength = state.gameHistory.length - 1;
      const latestMoveLength = state.gameHistory[gameHistoryLength].length;

      //  Don't undo any further if we're at the initial step for this move
      if (latestMoveLength <= 1) return state;

      //  Remove the last step from the most recent move for the game history
      const newGameHistory = cloneDeep([
        ...state.gameHistory.slice(0, gameHistoryLength),
        state.gameHistory[gameHistoryLength].slice(0, latestMoveLength - 1),
      ]);

      //  Get the previous step from the most recent move
      const newGameState = cloneDeep(state.gameHistory[gameHistoryLength][latestMoveLength - 2]);

      return {
        ...state,
        gameState: newGameState,
        gameHistory: newGameHistory,
      };
    }

    case UNDO_MOVE_FINISHED: {
      return {
        ...state,
        gameHistory: cloneDeep(state.gameHistory.slice(0, state.gameHistory.length - 1)),
        entitiesMoving: false,
      };
    }

    case SET_GAME_SPEED: {
      return {
        ...state,
        gameSpeed: action.gameSpeed,
      };
    }

    default: {
      return state;
    }
  }
}
