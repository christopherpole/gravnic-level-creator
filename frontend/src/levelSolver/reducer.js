import { EDIT_LEVEL } from 'levelPreview/actions';
import { UPDATE_TILE, RESET_GRID } from 'levelEditor/actions';
import { LOAD_LEVEL_CONFIRMED } from 'levelManager/actions';
import { FIND_QUICKEST_SOLUTION } from 'api/actions';
import { SOLVE_LEVEL } from './actions';

export const initialState = {
  solving: false,
  loading: false,
  loaded: false,
  solution: null,
  maxMoves: null,
  error: false,
};

export default function levelSolverReducer(state = initialState, action) {
  switch (action.type) {
    case SOLVE_LEVEL: {
      return {
        ...state,
        solving: true,
      };
    }

    case EDIT_LEVEL: {
      return {
        ...state,
        solving: false,
      };
    }

    case FIND_QUICKEST_SOLUTION.PENDING: {
      return {
        ...initialState,
        solving: state.solving,
        loading: true,
      };
    }

    case FIND_QUICKEST_SOLUTION.FULFILLED: {
      return {
        ...state,
        error: false,
        loading: false,
        loaded: true,
        solution: action.payload.solution,
        maxMoves: action.payload.maxMoves,
      };
    }

    case FIND_QUICKEST_SOLUTION.REJECTED: {
      return {
        ...state,
        error: true,
        loading: false,
      };
    }

    case LOAD_LEVEL_CONFIRMED: {
      return {
        ...state,
        solution: action.level.solution,
        maxMoves: action.level.maxMoves,
      };
    }

    case RESET_GRID:
    case UPDATE_TILE: {
      return initialState;
    }

    default: {
      return state;
    }
  }
}
