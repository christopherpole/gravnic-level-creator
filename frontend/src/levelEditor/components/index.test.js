import React from 'react';
import { configure, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { spy } from 'sinon';
import { ENTITIES } from 'gravnic-game';

import { initialState } from '../reducer';
import { LevelEditor, TileWrapper } from './index';

configure({ adapter: new Adapter() });

describe('<LevelEditor />', () => {
  let props;

  beforeEach(() => {
    props = {
      tiles: initialState.tiles,
      updateTileAction: () => {},
      startDragAction: () => {},
      stopDragAction: () => {},
    };
  });

  it('Renders without exploding', () => {
    const grid = shallow(<LevelEditor {...props} />);

    expect(grid).toHaveLength(1);
  });

  it('Matches the current snapshot', () => {
    const grid = shallow(<LevelEditor {...props} />);

    expect(toJson(grid)).toMatchSnapshot();
  });

  it('Fires the update tile action when clicking on a tile', () => {
    const tileClickSpy = spy();
    const grid = shallow(<LevelEditor {...props} updateTileAction={tileClickSpy} />);
    const tile = grid.find(TileWrapper).at(44);
    tile.simulate('click');

    expect(tileClickSpy.calledOnce).toBe(true);
    expect(tileClickSpy.calledWith(44)).toBe(true);
  });

  it('Fires start dragging action when the user begins and finishes dragging on a non-linked tile', () => {
    const startDragActionSpy = spy();
    const stopDragActionSpy = spy();
    const grid = shallow(
      <LevelEditor
        {...props}
        stopDragAction={stopDragActionSpy}
        startDragAction={startDragActionSpy}
      />,
    );
    const tilesWrapper = grid.find(TileWrapper).at(2);

    //  Start dragging
    expect(startDragActionSpy.calledOnce).toBe(false);
    tilesWrapper.simulate('mouseDown');
    expect(startDragActionSpy.calledOnce).toBe(true);
    expect(startDragActionSpy.calledWith(2)).toBe(true);
  });

  it(
    'Fires the update tile action if dragging over a tile that is not the same as the selectedtile',
  );

  it(
    'Does not fires the update tile action if dragging over a tile that is the same as the selectedtile',
  );
});
