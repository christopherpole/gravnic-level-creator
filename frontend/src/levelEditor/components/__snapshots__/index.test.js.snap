// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LevelEditor /> Matches the current snapshot 1`] = `
<styled.div
  id="editor-grid"
>
  <styled.div>
    <styled.div
      className="tile"
      key="0"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="1"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="2"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="3"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="4"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="5"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="6"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="7"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="8"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="9"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="10"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="11"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="12"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="13"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="14"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="15"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="16"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="17"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="18"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="19"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="20"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="21"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="22"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="23"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="24"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="25"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="26"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="27"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="28"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="29"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="30"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="31"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="32"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="33"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="34"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="35"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="36"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="37"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="38"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="39"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="40"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="41"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="42"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="43"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="44"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="45"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="46"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="47"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="48"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="49"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="50"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="51"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="52"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="53"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="54"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="55"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="56"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="57"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="58"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="59"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="60"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="61"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="62"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="63"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="64"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="65"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="66"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="67"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="68"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="69"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="70"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="71"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="72"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="73"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="74"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="75"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="76"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="77"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="78"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="79"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="80"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="81"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="82"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="83"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="84"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="85"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="86"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="87"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="88"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="89"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="90"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="91"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="92"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="93"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="94"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="95"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="96"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="97"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="98"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <styled.div
      className="tile"
      key="99"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
    >
      <Connect(Component)
        tileId="1"
      />
    </styled.div>
    <Connect(Component) />
  </styled.div>
</styled.div>
`;
